crates/sui-framework/packages/sui-framework/sources/kiosk/kiosk.move

```diff
@@ -412,6 +415,181 @@ module sui::kiosk {
         coin::take(&mut self.profits, amount, ctx)
     }
 
+    // === Kiosk Extensions API ===
+
+    /// Extension is trying to access a permissioned action while being disabled.
+    const EExtensionDisabled: u64 = 13;
+    /// Extension is trying to access a permissioned action while not having
+    /// the required permission.
+    const EExtensionNotAllowed: u64 = 14;
+    /// Extension is trying to access a permissioned action while not being
+    /// authorized to use the type.
+    const EExtensionNotAllowedForType: u64 = 15;
+
+    /// The Extension struct contains the data used by the extension and the
+    /// configuration for this extension. Stored under the `ExtensionKey`
+    /// dynamic field.
+    struct Extension has store {
+        /// Storage for the extension, an isolated Bag. By putting the extension
+        /// into a single dynamic field, we reduce the amount of fields on the
+        /// top level (eg items / listings) while giving extension developers
+        /// the ability to store any data they want.
+        storage: Bag,
+        /// Bitmask of permissions that the extension has (can be revoked any
+        /// moment). It's all or nothing policy - either the extension has the
+        /// required permissions or no permissions at all.
+        permissions: u32,
+        /// List of types that the extension is allowed to use. It only affects
+        /// permissioned actions (eg `place`, `lock`). The Kiosk owner can limit
+        /// an extensions functionality to a set of specified types.
+        types: vector<TypeName>,
+        /// Whether the requested permissions are still available. While the
+        /// owner can't delete the `Bag` storage unless the extension provides
+        /// the functionality, they can revoke the permissions at any time,
+        /// making the extension a harmless storage - similar to Kiosk having an
+        /// exposed UID.
+        is_enabled: bool,
+    }
+
+    /// The `ExtensionKey` is a typed dynamic field key used to store the
+    /// extension configuration and data. `Ext` is a phantom type that is used
+    /// to identify the extension witness.
+    struct ExtensionKey<phantom Ext> has store, copy, drop {}
```

Comment:
There's multiple reasons for having keys custom in Kiosk:

1. Easy to navigate - when I query dynamic fields I can detect extension because it starts with 0x2::kiosk::ExtensionKey, otherwise it would be a random field which may or may not mean an extension - bad devx.

2. Allowing third party keys on the base level would allow the third party read this fields avoiding the extensions functions. While it's not the case - `Extension` struct has no public getters; it's a general pattern which is coherent with other dynamic fields in the Kiosk (and there're many).